### PR TITLE
bpo-41076: Pre-feed the parser with the f-string expression location

### PR DESCRIPTION
**Pull request by lysnikolaou**
This PR changes the parsing of f-string expressions with the new
parser. The parser gets pre-fed with the location of the expression
itself (not the f-string, which was what we were doing before). This
allows us to completely skip the shifting of the AST nodes after
the parsing is completed.

https://bugs.python.org/issue41076

### CODE REVIEW

**Comment by pablogsal** [comment]
Wow, this is an excellent insight and looks fantastic! I can try to review it later tomorrow but I have an initial question:

Since we are storing the offset in fstring_col_offset, what happens with nested f-strings? Every parser instance has its own 'fstring_col_offset' and therefore they will not collide or there is something to consider on the propagation of these values?

File: Parser/string_parser.c [hunk 1]
```diff
@@ -598,11 +370,17 @@ fstring_compile_expr(Parser *p, const char *expr_start, const char *expr_end,
         return NULL;
     }
 
-    str[0] = '(';
+    str[0] = '{';
```
Comment (pablogsal) [inline Parser/string_parser.c:379]: Why the change to `{`? If I remember correctly, that parenthesis were to force parsing as expressions some constructs. Is this being preserved?

Note: I see that you are changing them to '(' after the call to `fstring_find_expr_location`. I think I understand why you need them when you go into `fstring_find_expr_location` but this is very subtle and certainly deserves a comment I think :)

Comment (lysnikolaou) [reply Parser/string_parser.c:379]: > I think I understand why you need them when you go into `fstring_find_expr_location`

This is due to the `strstr` call in https://github.com/python/cpython/blob/c96d00e88ead8f99bb6aa1357928ac4545d9287c/Parser/string_parser.c#L522. 

> this is very subtle and certainly deserves a comment I think :)

Good point! I'll add a comment.

**Comment by lysnikolaou** [comment]
> Since we are storing the offset in fstring_col_offset, what happens with nested f-strings? Every parser instance has its own 'fstring_col_offset' and therefore they will not collide or there is something to consider on the propagation of these values?

Nested f-strings should be handled correctly already. I checked `test_fstring.py` for tests the test the node locations of nested f-strings before pushing and there is one, so I guess it's ok. Hand-testing with some more examples verified this.

`fstring_col_offset` actually only gets used for the caret in the case of an error. Because only the expression part of the f-string is shown on an error, we need to shift the caret accordingly.

Nested f-strings are no problem because `t->col_offset` points to the *correct* column offset of the nested f-string (this is due to the pre-feeding of the parser with a beginning offset and due to correctly computing of the offset in `_PyPegen_fill_token` when it first gets fetched from the tokenizer). So, if the column offset of the nested fstring is correct, it only gets down to `fstring_find_expr_location` computing the correct amount of characters between the start of the nested f-string and the start of its expression. This function has been battle-tested by previous Python versions and in the old version of the new parser (the one before this PR), so I guess we're okay.

**Comment by pablogsal** [comment]
Although is not caused by this PR, I noticed that we have some inconsistency with the old parser regarding f-strings. For instance:

```
❯ ./python -X oldparser
Python 3.9.0b3+ (heads/deprecation_warnings-dirty:21b84dd204, Jun 21 2020, 19:23:13)
[GCC 10.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x =23; f'\n{1+2} - {lel $ 2}'
  File "<fstring>", line 1
    (lel $ 2)
         ^
SyntaxError: invalid syntax
>>>

```
>>> x =23; f'\n{1+2} - {lel $ 2}'
  File "<stdin>", line 1
    (lel $ 2)
             ^
SyntaxError: invalid syntax
>>>
```
```

**Comment by lysnikolaou** [comment]
Oh, this is a bug that needs to be fixed. I think I know what's causing it.

**Comment by lysnikolaou** [comment]
I realised that there was a bug in my implementation, in case the fstring wasn't starting at offset 0 (when we compute the offset of the caret in the error text, we also need to account for all the characters before the f-string, in case it's not at offset 0). The fix also fixes the bug @pablogsal mentioned above, so that's a win-win.

**Comment by pablogsal** [comment]
Closing and reopening to trigger the CI again

File: Parser/string_parser.c [hunk 2]
```diff
@@ -598,11 +370,21 @@ fstring_compile_expr(Parser *p, const char *expr_start, const char *expr_end,
         return NULL;
     }
 
-    str[0] = '(';
+    // Having curly braces at the beginning and at the end is needed so that the call
+    // to strstr in fstring_find_expr_location works correctly.
```
Comment (pablogsal) [inline Parser/string_parser.c]: Apologies if I am being too insistent on this, but I think it would be better if we clarify here what "works correctly" means, because I am afraid that I do understand the sentence because I already understand the problem, but for someone new to this and reading the code is unclear what the problem is or what "works correctly" means. (I talk also a higher level from just 'it will fail the sub-string match').

Comment (lysnikolaou) [reply Parser/string_parser.c]: No need to apologize. You're totally right that a newcomer would have to search around to understand this comment, which is at least suboptimal. I'll try to come up with something more helpful.

Comment (lysnikolaou) [reply Parser/string_parser.c]: @pablogsal How does the new version look?

Comment (pablogsal) [reply Parser/string_parser.c]: Fantastic!

**Review by pablogsal — APPROVED** [review]

**Comment by lysnikolaou** [comment]
🎉 🎉 

I feel like this is one of the most major improvements we've had lately. Thanks for the review @pablogsal!

**Comment by pablogsal** [comment]
> I feel like this is one of the most major improvements we've had lately. Thanks for the review @pablogsal!

Yeah, me too! This was a very good insight and a great job. Well done!
